### DISCUSSION TITLE
Mongoz allows to take database name at run-time.

### DISCUSSION THREAD

**Discussion by harshalizode** [dymmond/esmerald#278] [Q&A] [answered]
Hello,
In MongoDB, I have used multiple databases for each organisation, and every database has the same collections like user, employee, branch, etc. In the Mongoz documentation, they have provided the database name in the Meta class; however, this is not a mandatory field.If we can't pass the database name in the meta information, then how can we pass the database name at the query execution using Mongoz?

Thanks.

**Comment by tarsil** [comment]
@harshalizode isn't the database attribute in the Meta of MongoZ document not being enforced? Of not passed, raises an error? If not, then it's a bug that needs addressing in Mongoz.

In theory, the database attribute should be mandatory.

Thank you for this, of course.

**Reply by harshalizode** [reply, marked as answer]
Yes, it is mandatory in meta to pass the database name.
Thank you

**Reply by tarsil** [reply]
Not a problem. How are you finding Esmerald and Mongoz so far? Also, I hope I answered your question 👍🏼

**Reply by harshalizode** [reply]
Thank you for your quick response.
We have just started development with Esmerald framework, as we are having experience of Django framework, it is easy to understand all the concepts. So far we have found it is good framework for API development.
Regarding Mongoz, we are still exploring on it.

**Reply by tarsil** [reply]
Oh thank you @harshalizode . The feeling you are having with Esmerald is intentional. I myself came from years of Django and some other frameworks but Django feel was always what pleased me the most. Esmerald was designed to be futuristic for API development and applications with the latest features of python and maintaining the same Django level experience. I'm glad you all are enjoying. 

You are all welcome to also contribute if you see fit. Features, bugs, documentation, whatever that you can see that can make it better 👍🏼

**Comment by tarsil** [comment]
@harshalizode this will be also addressed today and maybe some improvements will be added to Mongoz.

Or we enforce this, or we allow to pass the dB name in the query but most likely we will enforce the database name as it was supposed to.

**Reply by harshalizode** [reply]
Okay, for now, is there any way to change the database while querying in MongoDB? Or we need to create a set_database method in the Mongoz manager to set the database before the query execution. If we can't pass the database name in the set_database method, then it will take the default database name that was provided in the Meta class.

**Reply by tarsil** [reply]
That is a good addition for a feature actually. We kinda do something similar with the multi tenancy using [Edgy](https://edgy.tarsild.io) and [Saffier](https://saffier.tarsild.io).

What do you say about you opening a PR with that and we can go from there? it would be great to have your contributions.

This change is in Mongoz (not Esmerald).

Maybe we could have something like this:

```python
User.objects.using("DB_NAME").all()
```

Where the using will be used instead of the database from meta.

**Reply by harshalizode** [reply]
Okay, I will work on it.

**Reply by tarsil** [reply]
You also got into my head in a good way. I started playing around to see how it could be done. Maybe next week something will come out 🙂

**Reply by harshalizode** [reply]
Hi,
How do we use the database in the manager's method to querying in documents?, so that I can adjust the logic appropriately for now.
Where is the database connection being stored?
Thanks.

**Reply by tarsil** [reply]
Hi, what database? Mongoz? Why don't you have a look at the documentation?

I'm struggling to understand the question, apologies for that.

https://mongoz.tarsild.io and this would probably help you since you can also see the source code 🙂

If you are referring to Mongoz, then the connection is stored inside the Registry:

https://github.com/tarsil/mongoz/blob/main/mongoz/core/connection/registry.py

**Reply by harshalizode** [reply]
Okay, I will check.
Thank you.

**Reply by harshalizode** [reply]
Hi,
Did you implement multi-tenancy for database selection at runtime?
Thank you.

**Reply by tarsil** [reply]
So, multi tenancy and NoSQL is something that is tricky to approach since the 3 Pilars of multi tenancy, at least so far, implies a relational database so, with Mongoz we didn't add that because of the nature of the DB but it can be thought and added, yes.

**Reply by harshalizode** [reply]
Hello, I have implemented the using() method into manager, i.e. `User.objects.using("DB_NAME").all()`. It works, but can we inject the database name via interceptors or middleware, such as Django queryset manager?
We don't want to enforce the database name while querying MongoDB.

**Reply by tarsil** [reply]
In MongoZ? If it works, nothing like opening a PR there and add that functionality + testing? It would be great to have that actually.

Regarding your question, The way you manipulate that via interceptors or middleware is up to you. I don't see the reason why not. One way to use that via those, would be probably by setting a context var in the middleware that can be then used and read anywhere else but that design and implementation would depend on your design.

**Reply by harshalizode** [reply]
Setting a context variable in the middleware? How did we define it?
It should be useful for all requests?, Because each requested user is in a different database, I'd like to configure it via middleware or interceptor rather than passing through a query.

**Reply by harshalizode** [reply]
For the time being, I've added the functionality; once the overall flow is complete, I'll test it and make any necessary changes. If everything works as expected, I'll let you know. I really appreciate your prompt reply. It is help me a lot to understand the concepts quickly. Should any obstacles arise, I will get in touch with you once again.

**Reply by tarsil** [reply]
That is understandable. One way to actually do something like this is having a table where you map a user with the corresponding database and then in the middleware you check the request for the user against that same database and map the query.

This is more multi tenancy design than technology itself and these are just ideas.

**Reply by tarsil** [reply]
Of course, I will try to help as much as I can and that functionality you did for MongoZ, you can always contribute by adding it in a PR against MongoZ as well.

It is more than welcomed.

**Reply by harshalizode** [reply]
Thank you.

**Reply by harshalizode** [reply]
Hello, I am attempting to create an employee document, But it failed to create it because of the abstract property in the Meta class.
The code is below.
```
 employee = await self.model_class.using(db).create(**kwargs)
 doc_id = employee.id
```
Document:
```
{
     'month_period': '202403', 
     'employee_code': '00190', 
     'full_name': 'Emp_Name_190', 
     'date_of_joining': '2011-10-31 00:00:00.000', 
     'e_mail': 'empname190@mail.com', 
     'mobile_no': '3222470621', 
     'domain_login_id': 'Emp_Name_190', 
     'line_of_business': 'Branch', 
     'last_modified_date': '2024-03-28 17:53:51.163'
}
```
Model:

```
class Employee(Document):

    month_period: str = mongoz.Integer
    full_name: str = mongoz.String
    employee_code: str = mongoz.String
    date_of_joining: str = mongoz.DateTime
    e_mail: str = mongoz.String
    mobile_no: str = mongoz.String
    domain_login_id: str = mongoz.String
    line_of_business: str = mongoz.String
    last_modified_date: str = mongoz.DateTime

    objects: ClassVar[CustomManger] = CustomManger()

    class Meta:
        abstract = False
        registry = registry
        database = default_db_name
```

Its raises:
```
raise AbstractDocumentError(
mongoz.exceptions.AbstractDocumentError: Employee(id=None) is an abstract class. This operation is not allowed
```
The problem is that it treats the abstract property default as True.
I set the abstract argument to False in the Meta class, and it still raises the same exception.

**Comment by harshalizode** [comment]
> Hi, what database? Mongoz? Why don't you have a look at the documentation?
> I am trying to implement changes in the registry and query manager using the "using" method, so I need a guidance to implement changes to make a connection with the mongoDB with the different databases at the run time .
> I'm struggling to understand the question, apologies for that.
> 
> https://mongoz.tarsild.io and this would probably help you since you can also see the source code 🙂
> 
> If you are referring to Mongoz, then the connection is stored inside the Registry:
> 
> https://github.com/tarsil/mongoz/blob/main/mongoz/core/connection/registry.py

**Reply by tarsil** [reply]
Apologies for my confusion. I understood after. MongoZ uses Motor to manage the connections inside the Registry. 

Mongoz follows the same patterns of Saffier and Edgy. More for consistency.

**Reply by harshalizode** [reply]
Thank you for providing the clarification regarding MongoZ and Motor, and I now understand how Motor manages the connections within the registry.

**Reply by tarsil** [reply]
No problem @harshalizode 🙂

**Comment by harshalizode** [comment]
Thank you for your response, waiting for the version 0.8.0. :D
